modules/node_modules/@ncigdc/components/AddCaseFilesToCartButton.js

```diff
@@ -0,0 +1,87 @@
+/* @flow */
+
+import React from 'react';
+import _ from 'lodash';
+import { connect } from 'react-redux';
+import { compose, withState, withProps, withHandlers } from 'recompose';
+
+import Dropdown from '@ncigdc/uikit/Dropdown';
+import ShoppingCartIcon from 'react-icons/lib/fa/shopping-cart';
+import TrashIcon from 'react-icons/lib/fa/trash';
+import { Tooltip } from '@ncigdc/uikit/Tooltip';
+import { addAllFilesInCart, removeFilesFromCart } from '@ncigdc/dux/cart';
+import withFilters from '@ncigdc/utils/withFilters';
+
+const styles = {
+  dropdownContainer: {
+    left: 0,
+    right: 'auto',
+  },
+};
+
+const AddCaseFilesToCartButton = compose(
+  connect(state => ({ cart: state.cart })),
+  withState('isLoading', 'setIsLoading', false),
+  withFilters(),
+  withProps(({ filters }) => ({
+    isFilteringOnFiles: filters && _.some(filters, (item) => {
+      const content = item.hasOwnProperty('content') ? item.content : item;
+      return content.field.indexOf('files.') === 0;
+    }),
+  })),
+  withHandlers({
+    handleDropdownActivate: ({ relay, setIsLoading, filters, isFilteringOnFiles, hasFiles }) => () => {
+      if (!hasFiles) {
+        // this case has no files to fetch.
+        return;
+      }
+      setIsLoading(true);
+      relay.setVariables({
+        isFileDataRequired: true,
+        filesFilters: isFilteringOnFiles ? filters : null,
+      }, (readyState) => _.some([readyState.ready, readyState.aborted, readyState.error]) && setIsLoading(false));
+    },
+  })
+)(({ hasFiles, files = [], filteredFiles = [], dispatch, handleDropdownActivate, isLoading, isFilteringOnFiles }) => (
+  <Dropdown
+    dropdownStyle={styles.dropdownContainer}
+    dropdownClassName={isLoading ? 'hidden' : 'dropdown-menu'}
+    isDisabled={!hasFiles}
+    button={
+      hasFiles
+        ? (
+          <button className="btn btn-default dropdown-toggle fa fa-shopping-cart" style={{ padding: '0 4px' }}>
+            <span className={isLoading ? 'fa fa-spinner fa-spin' : 'caret'} style={{ marginLeft: '0.5rem' }} />
+            <span className="icon-btn-label">Case Actions</span>
+          </button>
+        )
+        : (
+          <Tooltip Component={<span>This case has no files.</span>}>
+            <button className="btn disabled fa fa-shopping-cart" style={{ padding: '0px 17px 0 4px' }} />
+          </Tooltip>
+        )
+    }
+    onActivate={handleDropdownActivate}
+  >
+    {
+      isLoading
+        ? <li> Loading case files... </li>
+        : ([
+          !!files.length && <li onClick={() => dispatch(addAllFilesInCart(files))} aria-label="Add all Case files to the Cart" role="button" key="addAll">
+            <ShoppingCartIcon style={{ marginRight: '1em' }} />Add all Case files to the Cart ({files.length})
+          </li>,
+          !!files.length && <li onClick={() => dispatch(removeFilesFromCart(files))} aria-label="Remove all Case files from the Cart" role="button" key="removeAll">
```

Comment:
>As I mentioned earlier I am suggesting using `toggleFilesInCart` only to remove files.

I understand, that's where I'm saing the issue lies. Let me try to rephrase

When I'm trying to remove files from a cart, with a call to 
`toggleFilesInCart([1, 2, 3])` 
I have no guarantee that the cart will no longer contain the files I passed in.
Whether [1, 2, 3] is in the cart after the call depends on whether [1, 2, 3] was in the cart before the call. If a file is passed in to the function that wasn't already in the cart, it would be added to the cart. Places that use the function would also have to guard against multiple invocations. The second time that function gets called it becomes an "add" instead of a "remove"

This kind of unexpected behavior of  "wanting to only remove files" but the result of the call to `toggleFilesInCart` actually adding files, is why the `removeFilesFromCart` function was added - so that there is a guarantee that when there is a desire to remove files, under no circumstances would files be added to the cart.

The defining logic between `toggleFilesInCart` and `removeFilesFromCart` is that one is an "xor", the other is a "difference". Sometimes the xor results in a difference. Other times it results in an addition. (places that are currently using `addAllFilesInCart` (which should perhaps be renamed to `addFilesToCart`) could also be changed to use`toggleFilesInCart` with similar drawbacks)

The only instance where an xor on the cart by an array of passed in files would be useful would be if we desire to have a function that we call, passing in an array of files, adding the ones that don't exist and removing the ones that do. 

```
cart = [1, 2, 3];

toggleFilesInCart([1, 2, 3, 4, 5, 6]);

// cart is now [4, 5, 6];
```